keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +337,302 @@ or later. The dual-stack feature requires no changes to this API.
+  3. Set `spec.preferDualStack: true` and set `spec.ClusterIPs` to max of one ip. This ip will be treated as the primary ip for the resultant service (if available for allocation). api-server will default `spec.ipFamilies` and set `spec.ClusterIPs: ['ip provided by user', 'allocated ip from cluster secondary ip family']`
+
 
-Dual-stack services (having both an IPv4 and IPv6 cluster IP) are currently out of scope of this proposal. Should there be use-cases where dual-stack Services are needed then we can revisit.
+The below are sample services that demonstrate the above rules
+
+```yaml
+# dual stack service (required)
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  preferDualStack: true # set by user
+  ipFamilies: # set by user
+    - IPv4
+    - IPv6
+  clusterIP: 1.2.3.4  # allocated
+  clusterIPs: # allocated
+    - 1.2.3.4
+    - 2001::1
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+
+# dual stack service (required)
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  preferDualStack: true # defaulted
+  ipFamilies: # defaulted
+    - IPv6
+    - IPv4
+  clusterIP: 2001::1  # set by user
+  clusterIPs: # set by user
+    - 2001::1
+    - 1.2.3.4
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+
+# dual stack service (optional)
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  preferDualStack: true # set by user 
+  ipFamilies:
+    - IPv6 # defaulted, cluster does not support dual stack services, only one ipv6 cidr is provided for services
+  clusterIP: 2001::1  # allocated
+  clusterIPs: # allocated
+    - 2001::1 # note single entry
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+
+# dual stack service (optional)
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  preferDualStack: true # set by user 
+  ipFamilies:
+    - IPv6 # set by user
+    - IPv4 # defaulted (cluster supports dual stack)
+  clusterIP: 2001::1  # allocated
+  clusterIPs: # allocated
+    - 2001::1
+    - 1.2.3.4
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+
+```
+
+##### Notes on NodePort Allocation
+
+NodePort reservation will be constant across both families. That mean a reservation of `NodePort: 12345` will happen across both families irrespective of the ip family of the service.
 
 #### Type Headless services
 
-\<TBD\>
+1. For Headless with selector, the value of `spec.preferDualStack` and `spec.ipFamilies` will drive the endpoint selection. Endpoint slice controller will select endpoints with ip families that match the value of these two flags.
```

Comment:
Actually, only `ipFamilies` matters to controllers.  Nobody except the service allocator should have to look at the `prefer` flag, right?

Reply:
> What we should do now is boil it down to a smaller form (slides?) and get an outside API reviewer to read it. I can sign up to condense it if you want, Kal

@thockin yes please condense. I will attend the api review as well.

Reply:
> Actually, only `ipFamilies` matters to controllers. Nobody except the service allocator should have to look at the `prefer` flag, right?

Yes. Because there will be cases where user wants dualstack `preferDualStack: true` but won't get it, and `IPFamilies` is the old field that reflect that fact (and `ClusterIPs` but that will require testing ips for families wherever controller is concerned)

Reply:
But the only reason a service would `preferDualStack` but not get it is because the cluster isn't dual-stack. Or at least "doesn't have dual-stack service CIDRs". I suppose in theory we could support "I want a cluster with dual-stack pods and dual-stack headless services but only single-stack ClusterIP services" but that's weird and I think it's better to say that nobody except the service allocator looks at `preferDualStack`

Reply:
This is a relatively minor point but `ipFamiles[]` has all the information it needs.  Eg.

time +0: user creates a headless service with `prefer=true`, but cluster is single-stack.  They get `ipFamiles = [ v4 ]`.

time +1: endpoint controllers see `ipFamiles = [ v4 ]` and only consider v4 addresses

time +2: admin upgrades cluster to dual-stack

time +3: endpoint controllers see `ipFamiles = [ v4 ]` and only consider v4 addresses

The point being that the controllers should NOT change at T+3, because the service pre-existed dual-stack enablement.